FaunaDB.Client/Encoding/Encoder.cs

```diff
@@ -0,0 +1,244 @@
+﻿using System;
+using System.Runtime.CompilerServices;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using FaunaDB.Query;
+using FaunaDB.Types;
+using FaunaDB.Attributes;
+using System.Linq.Expressions;
+
+namespace FaunaDB.Encoding
+{
+    public static class Encoder
+    {
+        public static Expr Encode(object obj) =>
+            new EncoderImpl().Encode(obj);
+    }
+
+    class EncoderImpl
+    {
+        static readonly Dictionary<Type, Func<object, Expr>> converters = new Dictionary<Type, Func<object, Expr>>();
+
+        Stack<object> stack = new Stack<object>();
+
+        public Expr Encode(object obj)
+        {
+            if (obj == null)
+                return NullV.Instance;
+
+            if (typeof(Expr).IsInstanceOfType(obj))
+                return (Expr)obj;
+
+            if (stack.Contains(obj))
```

Comment:
How does `Contains` check for equality? Do we have the risk of haring object pools returning `true` for equal objects but at different pointers? Similar to comparing strings in java with `==`?